<a href="https://colab.research.google.com/github/cam2149/eda_couse/blob/Unidad1/Impacto_nulos_performance.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Impacto de los nulos en el performance

## Importamos las librerias

In [2]:
#Importamos las librerias que usaremos en este practica
## Pandas es una librería de Python especializada en la manipulación y el análisis de datos. Ofrece estructuras de datos
# y operaciones para manipular tablas numéricas y series temporales, es como el Excel de Python
import pandas as pd
## NumPy es una biblioteca para el lenguaje de programación Python que da soporte para crear vectores y matrices grandes
#multidimensionales, junto con una gran colección de funciones matemáticas de alto nivel para operar con ellas
import numpy as np
## Seaborn ofrece una variedad de potentes herramientas para la visualización de datos, incluyendo gráficos de dispersión,
#gráficos de líneas, gráficos de barras, mapas de calor, y muchos más
import seaborn as sns

sns.set_theme(style="whitegrid")

## 1. Cargar el data set

# Usando google colab podemos cargar el data set de dos formas, podemos usar la libreria de drive de google o podemos subir el archivo al colab.

### 1.1 Usando la libreria de drive

In [ ]:
#Importamos la libreria de drive



In [ ]:
#Importamos la libreria de drive para conectarnos con el Drive de nuestra cuenta Google
from google.colab import drive
drive.mount('/content/drive', force_remount=True)
# En esta oportunidad no se va hacer, se cargara el archivo desde Github

Mounted at /content/drive


In [ ]:
# Cople el archivo a su Gdrive desde: https://raw.githubusercontent.com/armandoordonez/eda_couse/main/data/diabetes-NAN.csv

### 1.2 Cargando el archivo a google colab


In [3]:
# Modificamos la ruta de acuerdo a su Gdrive
#Usamos pd (la libreria de pandas, que fue definida en la celda donde importamos las librerias) para leer el dataset y asignarlo a df
df = pd.read_csv('https://raw.githubusercontent.com/cam2149/eda_couse/main/data/diabetes-NAN.csv',    #Ruta completa a la ubicación del archivo de interés.
                 na_values='?')              #Como el archivo contiene algunos caracteres ?, con este comando pandas los interpreta como nan (not a number).

In [8]:
# Mostrar las primeras 5 filas del dataset utilizando el método dataframe.head()
print("Las primeras 5 filas del data frame:")
df.head(5)

Las primeras 5 filas del data frame:


Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0          6.0    148.0           72.0           35.0      0.0  33.6   
1          1.0     85.0           66.0           29.0      0.0  26.6   
2          8.0      NaN           64.0            0.0      0.0  23.3   
3          1.0     89.0           66.0           23.0     94.0  28.1   
4          0.0      NaN            NaN           35.0    168.0  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [9]:
# Crear la lista de nombres para el encabezado
headers = ["_Pregnancies","_Glucose",	"_BloodPressure",	"_SkinThickness",	"_Insulin",	"_BMI",	"_DiabetesPedigreeFunction","_Age","_Outcome"]
print("encabezados\n", headers)

encabezados
 ['_Pregnancies', '_Glucose', '_BloodPressure', '_SkinThickness', '_Insulin', '_BMI', '_DiabetesPedigreeFunction', '_Age', '_Outcome']


In [10]:
#Reemplazamos el encabezado y verificamos nuestro data frame:

df.columns = headers
df.head(10)

_Pregnancies  _Glucose  _BloodPressure  _SkinThickness  _Insulin  _BMI  \
0           6.0     148.0            72.0            35.0       0.0  33.6   
1           1.0      85.0            66.0            29.0       0.0  26.6   
2           8.0       NaN            64.0             0.0       0.0  23.3   
3           1.0      89.0            66.0            23.0      94.0  28.1   
4           0.0       NaN             NaN            35.0     168.0  43.1   
5           5.0     116.0            74.0             NaN       0.0  25.6   
6           3.0      78.0            50.0            32.0      88.0  31.0   
7           NaN     115.0             0.0             0.0       NaN  35.3   
8           2.0     197.0            70.0            45.0     543.0  30.5   
9           8.0     125.0            96.0             0.0       NaN   0.0   

   _DiabetesPedigreeFunction  _Age  _Outcome  
0                      0.627    50         1  
1                      0.351    31         0  
2                      0.672    32         1  
3                      0.167    21         0  
4                      2.288    33         1  
5                      0.201    30         0  
6                      0.248    26         1  
7                      0.134    29         0  
8                      0.158    53         1  
9                      0.232    54         1

## 2. Observaciones basicas del dataset

### Tamaño del dataset

In [15]:
df.shape

(768, 9)

### información general

Podemos observar el nombre de las columnas, si tienen nulos y cual es su tipo de dato

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   _Pregnancies               766 non-null    float64
 1   _Glucose                   759 non-null    float64
 2   _BloodPressure             761 non-null    float64
 3   _SkinThickness             759 non-null    float64
 4   _Insulin                   762 non-null    float64
 5   _BMI                       761 non-null    float64
 6   _DiabetesPedigreeFunction  766 non-null    float64
 7   _Age                       768 non-null    int64  
 8   _Outcome                   768 non-null    int64  
dtypes: float64(7), int64(2)
memory usage: 54.1 KB


In [18]:
df.head(10)

_Pregnancies  _Glucose  _BloodPressure  _SkinThickness  _Insulin  _BMI  \
0           6.0     148.0            72.0            35.0       0.0  33.6   
1           1.0      85.0            66.0            29.0       0.0  26.6   
2           8.0       NaN            64.0             0.0       0.0  23.3   
3           1.0      89.0            66.0            23.0      94.0  28.1   
4           0.0       NaN             NaN            35.0     168.0  43.1   
5           5.0     116.0            74.0             NaN       0.0  25.6   
6           3.0      78.0            50.0            32.0      88.0  31.0   
7           NaN     115.0             0.0             0.0       NaN  35.3   
8           2.0     197.0            70.0            45.0     543.0  30.5   
9           8.0     125.0            96.0             0.0       NaN   0.0   

   _DiabetesPedigreeFunction  _Age  _Outcome  
0                      0.627    50         1  
1                      0.351    31         0  
2                      0.672    32         1  
3                      0.167    21         0  
4                      2.288    33         1  
5                      0.201    30         0  
6                      0.248    26         1  
7                      0.134    29         0  
8                      0.158    53         1  
9                      0.232    54         1

Gracias a esta información, podemos ver que tenemos una pequeña cantidad de datos nulos

### Vista preliminar

Observamos los primeros y ultimos 5 datos del data set

In [19]:
df.head()

_Pregnancies  _Glucose  _BloodPressure  _SkinThickness  _Insulin  _BMI  \
0           6.0     148.0            72.0            35.0       0.0  33.6   
1           1.0      85.0            66.0            29.0       0.0  26.6   
2           8.0       NaN            64.0             0.0       0.0  23.3   
3           1.0      89.0            66.0            23.0      94.0  28.1   
4           0.0       NaN             NaN            35.0     168.0  43.1   

   _DiabetesPedigreeFunction  _Age  _Outcome  
0                      0.627    50         1  
1                      0.351    31         0  
2                      0.672    32         1  
3                      0.167    21         0  
4                      2.288    33         1

In [20]:
df.tail()

_Pregnancies  _Glucose  _BloodPressure  _SkinThickness  _Insulin  _BMI  \
763          10.0     101.0            76.0            48.0     180.0  32.9   
764           2.0     122.0            70.0            27.0       0.0  36.8   
765           5.0     121.0            72.0            23.0     112.0  26.2   
766           1.0     126.0            60.0             0.0       0.0  30.1   
767           1.0      93.0            70.0            31.0       0.0  30.4   

     _DiabetesPedigreeFunction  _Age  _Outcome  
763                      0.171    63         0  
764                      0.340    27         0  
765                      0.245    30         0  
766                      0.349    47         1  
767                      0.315    23         0

In [21]:
df._Pregnancies.describe()

count    766.000000
mean       3.828982
std        3.359232
min        0.000000
25%        1.000000
50%        3.000000
75%        6.000000
max       17.000000
Name: _Pregnancies, dtype: float64

# 3.1 Clasificación con datos nulos

In [22]:
#Escojamos nuestras variables predictoras
predictores = ['_Pregnancies','_Glucose','_BloodPressure','_SkinThickness','_Insulin','_BMI','_DiabetesPedigreeFunction','_Age']
#Escojamos que es lo que queremos encontrar
objetivo= '_Outcome'

#Recordemos que en nuestra variable Y va lo que queremos encontrar/predecir.
#Mientras en la vairable X van las variables que usaremos para encontrar a Y
y= df[objetivo].values
X= df[predictores].values

In [23]:
#Scikit-learn, also known as sklearn, is an open-source, machine learning and data modeling library for Python
from sklearn.model_selection import train_test_split #Split arrays or matrices into random train and test subsets.
# Crear conjuntos de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

(614, 8)
(154, 8)
(614,)
(154,)


Si intentamos correr este codigo, veremos que la regresión logisitica no puede lidear con valores nulos, así que usaremos otro modelo.

In [26]:
# Se importa el modelo
from sklearn.linear_model import LogisticRegression

#Se crea el modelo
log_reg = LogisticRegression(solver='lbfgs',max_iter=100)
#se entrena el modelo
log_reg.fit(X_train, y_train)

ValueError: Input X contains NaN.
LogisticRegression does not accept missing values encoded as NaN natively. For supervised learning, you might want to consider sklearn.ensemble.HistGradientBoostingClassifier and Regressor which accept missing values encoded as NaNs natively. Alternatively, it is possible to preprocess the data, for instance by using an imputer transformer in a pipeline or drop samples with missing values. See https://scikit-learn.org/stable/modules/impute.html You can find a list of all estimators that handle NaN values at the following page: https://scikit-learn.org/stable/modules/impute.html#estimators-that-handle-nan-values

In [25]:

# LogisticRegression does not accept missing values encoded as NaN natively. For supervised learning, you might want to consider sklearn.ensemble.HistGradientBoostingClassifier and Regressor which accept missing values encoded as NaNs natively. Alternatively, it is possible to preprocess the data, for instance by using an imputer transformer in a pipeline or drop samples with missing values. See https://scikit-learn.org/stable/modules/impute.html You can find a list of all estimators that handle NaN values at the following page: https://scikit-learn.org/stable/modules/impute.html#estimators-that-handle-nan-values
# Usamos otro modelo

from sklearn.ensemble import (HistGradientBoostingClassifier)

hgbm = HistGradientBoostingClassifier(random_state=42)
hgbm.fit(X_train, y_train)
hgbm.score(X_test, y_test)

0.7272727272727273

Nuestro resultado fue de 0.72, que no esta nada mal, ahora veamos que pasa si nos encargamos de los valores nulos

# 3.2 Clasificación sin datos nulos

In [27]:
#veamos cuantos nan tenemos en columnas
df.isna().sum()

_Pregnancies                 2
_Glucose                     9
_BloodPressure               7
_SkinThickness               9
_Insulin                     6
_BMI                         7
_DiabetesPedigreeFunction    2
_Age                         0
_Outcome                     0
dtype: int64

In [28]:
#veamos cuantos nan tenemos en total
df.isna().sum().sum()

42

Podemos ver, que solo 42 de las 768 observaciones tienen valores nulos, así que en este caso borraremos esos datos

In [29]:
df_clean=df.dropna()


In [30]:
df_clean.isna().sum()

_Pregnancies                 0
_Glucose                     0
_BloodPressure               0
_SkinThickness               0
_Insulin                     0
_BMI                         0
_DiabetesPedigreeFunction    0
_Age                         0
_Outcome                     0
dtype: int64

Ahora que tenemos un dataset limpio, repetimos el proceso

In [31]:
#Escojamos nuestras variables predictoras
predictores = ['_Pregnancies','_Glucose','_BloodPressure','_SkinThickness','_Insulin','_BMI','_DiabetesPedigreeFunction','_Age']
#Escojamos que es lo que queremos encontrar
objetivo= '_Outcome'

#Recordemos que en nuestra variable Y va lo que queremos encontrar/predecir.
#Mientras en la vairable X van las variables que usaremos para encontrar a Y
y= df_clean[objetivo].values
X= df_clean[predictores].values

In [32]:
# Crear conjuntos de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

(595, 8)
(149, 8)
(595,)
(149,)


In [33]:
#Se crea el modelo
log_reg = LogisticRegression(solver='lbfgs',max_iter=1000)
#se entrena el modelo
log_reg.fit(X_train, y_train)

LogisticRegression(max_iter=1000)

In [34]:
log_reg.score(X_test, y_test)

0.8187919463087249

Se puede observar que ya obtuvimos un mejor resultado sin nulos usando regresión logistica.

Dicho esto, usaremos el hgbm de nuevo para una mejor comparación

In [35]:
hgbm = HistGradientBoostingClassifier(random_state=42)
hgbm.fit(X_train, y_train)
hgbm.score(X_test, y_test)

0.7919463087248322

Como se puede ver, este modelo tambien presenta una mejora en su score.

In [ ]:
# Ejercicio: Utiliza otros modelos que soporten nulos para la predicción y evalua su score